Refactor Travis build to run "large" integration tests in a separate build stage

smurching opened the pull request:
## Summary of changes
This PR:
* Refactors tests into small & large stages (such that large tests only run after small tests pass).
* Splits existing .travis.yml logic into a number of CI scripts into the `travis` directory. This helps ensure that our contributor guide instructions for running tests (in `CONTRIBUTING.rst`) stay up to date with Travis by running the same bash test scripts in both.

## Goals of this PR
While the above changes won't eliminate test flakiness, they will:

* Generally speed CI under the assumption that most PR code iteration involves breaking & fixing small tests (i.e. large tests will run fewer times & so consume less Travis executor time)
* Make it clearer when a build fails due to flaky large tests
* Make it easier/faster to retrigger flaky tests (it'll be possible to rerun just large tests, instead of needing to run both the small & large tests).

If we find that large tests generally pass after small tests succeed, we might consider allowing PRs to merge after small tests pass & running large tests less frequently.

## Example builds
Compare [this master build](https://travis-ci.org/mlflow/mlflow/builds/517456525) to [this successful  PR build](https://travis-ci.org/mlflow/mlflow/builds/517591722). The total build time is actually ~10% worse after this PR (due to the serial nature of running large tests after small ones), but note that the small build completes in ~7 min, so any PR builds that fail during small tests (e.g. [this one](https://travis-ci.org/mlflow/mlflow/builds/517568177?utm_source=github_status&utm_medium=notification)) will take much less executor time.

## How is this PR tested?
Manually tested by running tests on the PR itself. Also verified manually that the updated contributor instructions (in CONTRIBUTING.rst) work.

smurching committed:
Run large & non-large tests via separate invocations to measure timings on Travis

smurching committed:
Merge master

smurching committed:
Try using stages?

smurching committed:
Add back language as Python

smurching committed:
Tweaks - make shell scripts runnablee

smurching committed:
Try to reduce number of Python builds, make install steps standard

smurching committed:
Fix typos in pytest

smurching committed:
[test-large] test out build trigger by commit message

smurching committed:
Unconditionally run large builds if small ones pass. Also try to fix Python version issue

smurching committed:
Remove backslashes that break large test command

smurching committed:
Try to reduce # of builds by removing first stage annotation as per https://github.com/travis-ci/beta-features/issues/42\#issuecomment-361732543

smurching committed:
Remove top level stage annotation?

smurching committed:
Try adding python key

smurching committed:
Try again

smurching committed:
Copy some code around

smurching committed:
Debug

smurching committed:
Remove stage again?

smurching committed:
Try once more to remove extra job...

smurching committed:
Give up on removing extra python job :(

smurching committed:
Make build script gross but I think things might work??

smurching committed:
Actually add large tests

smurching committed:
source install script

smurching committed:
Add large-only to exclusively run large tests

smurching committed:
Update contributor docs & bash scripts to print nicer output

smurching committed:
Update instructions

smurching committed:
Don't permanently set -ex. Also try disabling top-level build?

smurching committed:
Add back install: skip

smurching committed:
Try splitting up large & small Python test deps

smurching committed:
Add back torch, torchvision deps

smurching committed:
Add back TF, Pyspark deps

smurching committed:
Try getting rid of extra build stage, I think this time it'll actually work whoo

smurching committed:
Fix yaml

smurching committed:
Add back store deps to large tests & add logic to test scripts to fail if any commands exit with non-zero status

smurching committed:
Split lint out

smurching committed:
Fix check

smurching committed:
Fix script

smurching committed:
Try again...

smurching committed:
Move pytest out of linter deps

smurching committed:
Install typing to fix Python 2.7 torch build

smurching committed:
Refactor dep script a bit

smurching committed:
Debug yaml

smurching committed:
Name builds so they don't have a weird 'Python 3.6' prefix

smurching committed:
mkdir before attempting to download stuff into it

smurching committed:
Refactor scripts under travis directory

smurching committed:
Copy Tomas's R fix + fix small & large scripts for local use

smurching committed:
Fix install script

smurching committed:
Fix .travis.yml again

smurching commented on .travis.yml:
```diff
@@ -28,81 +46,45 @@ matrix:
+# Travis runs an extra top-level job for each build stage - depending on the build stage, we either
+# run small or large Python tests below.
 install:
-  - sudo mkdir -p /travis-install
```
These common build steps have been refactored into `./travis/install-common-deps.sh`

smurching commented on .travis.yml:
```diff
@@ -28,81 +46,45 @@ matrix:
-  # Install protoc
-  - wget https://github.com/google/protobuf/releases/download/v3.6.0/protoc-3.6.0-linux-x86_64.zip -O /travis-install/protoc.zip
-  - sudo unzip /travis-install/protoc.zip -d /usr
 script:
```
These Python test steps have been refactored into `./travis/run-small-python-tests.sh` & `./travis/run-large-python-tests.sh`

smurching commented on conftest.py:
```diff
@@ -1,4 +1,6 @@
 def pytest_addoption(parser):
+    parser.addoption('--large-only', action='store_true', dest="large_only",
```
Add a `--large-only` option so that we can exclusively run tests marked as large in `./travis/run-large-python-tests.sh`

smurching commented on test-requirements.txt:
```diff
@@ -1,32 +1,3 @@
-cryptography==2.5.0
-# test plugin
-tests/resources/mlflow-test-plugin/
+-r travis/small-requirements.txt
```
Make `test-requirements.txt` a pointer to `small-requirements.txt`, `large-requirements.txt`, etc. It should still work as it did before / install the same set of dependencies.

Splitting up the requirements files speeds up running small Python tests by avoiding installing large dependencies like TensorFlow & PySpark (these take ~4 min to install).

smurching commented on travis/install-common-deps.sh:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
Copied from the old .travis.yml `install` section ([see comment](https://github.com/mlflow/mlflow/pull/1097/files#r273299584))

smurching commented on travis/run-large-python-tests.sh:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
```
Split out from the old .travis.yml `script` section (compare to [here](https://github.com/mlflow/mlflow/pull/1097/files#diff-354f30a63fb0907d4ad57269548329e3L93))

smurching committed:
Debug

smurching committed:
Actually install correct deps

aarondav commented on .travis.yml:
```diff
@@ -1,22 +1,40 @@
 sudo: true
 language: python
+# Note: can remove this once Travis's default Python version switches to 3.6 on April 16, 2019.
```
This date has passed

aarondav commented on CONTRIBUTING.rst:
```diff
@@ -114,6 +114,38 @@ When developing, you can make Python changes available in R by running (from mlf
+    pytest tests/pyfunc --large
+
+Note: Certain model tests are not well-isolated (can result in OOMs when run in the same Python
+process), so simply invoking ``pytest`` or ``pytest tests`` is not known to work. If you'd like to
```
Maybe "is not known" --> "may not"

aarondav commented on CONTRIBUTING.rst:
```diff
@@ -123,20 +155,6 @@ Alternatively, you can generate the necessary files in ``mlflow/server/js/build`
-    pytest
-    ./lint.sh
-
-When running ``pytest --requires-ssh`` it is necessary that passwordless SSH access to localhost
```
Is this no longer relevant?

smurching replied:
Ah no it still is, I'll add it back + clarify that some tests aren't run in Travis by default (& how to run them).

aarondav commented on travis/run-small-python-tests.sh:
```diff
@@ -0,0 +1,12 @@
+err=0
+trap 'err=1' ERR
+
+pytest --cov=mlflow --verbose --ignore=tests/h2o --ignore=tests/keras \
```
Do we need to duplicate these `--ignore` calls? Couldn't we just mark these as large?

smurching replied:
Yep true - unfortunately there doesn't seem to be a great way to mark the specific test files/modules as large & it's a little heavyweight to put @pytest.mark.large before each large test, but it is helpful to make the usage of @pytest.mark.large consistent.

smurching committed:
Merge branch 'master' into travis-refactor

smurching committed:
* Add back instructions on how to run tests marked @pytest.mark.requires_ssh, @pytest.mark.release
* Mark all models/deployment tests as large so that we can more simply run small tests via 'pytest tests' (rather than requiring a bunch of --ignore directives)

aarondav committed:
Lint, pyarrow, and large tests

aarondav commented:
@smurching Pushed a commit which fixes lint errors, adds pyarrow to the small-test-requirements.txt, and I think fixed the changes to run-small-tests and run-large-tests which I think you swapped.

smurching committed:
Add back --ignore directives in small tests, because otherwise pytest tries to import
those modules. Also, explain how to add new small/large Python tests in contributor guide

smurching commented:
@aarondav thanks! Unfortunately it looks like the --ignore directives in the small tests (which we discussed in [this thread](https://github.com/mlflow/mlflow/pull/1097#discussion_r279048737)) are necessary - otherwise pytest tries to import the `test_keras_xxxx` etc modules during small tests, which fails as dependencies for MLflow model tests aren't installed in our small test runner. I added the `--ignore` back & also pushed a small addition to the contributor guide explaining how to add new small/large tests & corresponding dependencies.

smurching committed:
Remove unwanted change to PR template (will include in separate PR)

smurching commented on CONTRIBUTING.rst:
```diff
@@ -3,8 +3,6 @@ Contributing to MLflow
 We welcome community contributions to MLflow. This page describes how to develop/test your changes
 to MLflow locally.
 
-Python
```
Removed this because there's a separate section also titled "Python" (with analogous "Java" and "R" sections)  discussing the dev workflow for the Python package further below

smurching committed:
Merge branch 'master' into travis-refactor

aarondav committed:
Update run-small-python-tests.sh

aarondav committed:
Update run-large-python-tests.sh

aarondav committed:
Update run-large-python-tests.sh
